hsds/link_sn.py

```diff
@@ -236,99 +230,232 @@ async def PUT_Link(request):
+
+    # construct an item list from titles and group_ids
+    items = {}
+    if group_ids is None:
+        if not req_id:
+            msg = "no object id in request"
+            log.warn(msg)
+            raise HTTPBadRequest(reason=msg)
+        items[req_id] = titles
+    elif isinstance(group_ids, list):
+        if titles is None:
+            msg = "no titles - will return all links for each object"
+            log.debug(msg)
+        for group_id in group_ids:
+            items[group_id] = None
+    elif isinstance(group_ids, dict):
+        if titles is not None:
+            msg = "titles must not be provided if obj_ids is a dict"
+            log.warn(msg)
+            raise HTTPBadRequest(reason=msg)
+        for group_id in group_ids:
+            names_for_id = group_ids[group_id]
+            if not isinstance(names_for_id, list):
+                msg = "expected list of titles"
+                log.warn(msg)
+                raise HTTPBadRequest(reason=msg)
+            items[group_id] = names_for_id
+
+    log.debug(f"POST Links items: {items}")
+
+    # do a check that everything is as it should with the item list
+    for group_id in items:
+        if not isValidUuid(group_id, obj_class="Group"):
+            msg = f"Invalid group id: {group_id}"
+            log.warn(msg)
+
+        titles = items[group_id]
+
+        if titles is None:
+            log.debug(f"getting all links for {group_id}")
+        elif isinstance(titles, list):
+            for title in titles:
+                validateLinkName(title)  # raises HTTPBadRequest if invalid
+        else:
+            msg = f"expected list for titles but got: {type(titles)}"
+            log.warn(msg)
+            raise HTTPBadRequest(reason=msg)
+
+    username, pswd = getUserPasswordFromRequest(request)
+    if username is None and app["allow_noauth"]:
+        username = "default"
+    else:
+        await validateUserPassword(app, username, pswd)
+
+    domain = getDomainFromRequest(request)
+    if not isValidDomain(domain):
+        msg = f"Invalid domain value: {domain}"
+        log.warn(msg)
+        raise HTTPBadRequest(reason=msg)
+    bucket = getBucketForDomain(domain)
+
+    # get domain JSON
+    domain_json = await getDomainJson(app, domain)
+    verifyRoot(domain_json)
+
+    # TBD - verify that the obj_id belongs to the given domain
+    await validateAction(app, domain, req_id, username, "read")
+
+    resp_json = {}
+
+    if len(items) == 0:
+        msg = "no group ids specified for POST Links"
+        log.warn(msg)
+        raise HTTPBadRequest(reason=msg)
+    elif len(items) == 1 and not follow_links:
+        # just make a request to the datanode
+        group_id = list(items.keys())[0]
+        titles = items[group_id]
+        links = await getLinks(app, group_id, titles=titles, bucket=bucket)
+
+        resp_json["links"] = links
+    else:
+        # get multi obj
+        # don't follow links for the groups we visit!
```

Comment:
This comment seems misleading, since domain crawlers can follow links based on the parameter in `domain_crawl.py/get_links`.

Reply:
Fixed